core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -147,20 +139,28 @@ protected void doExecute(Task task, Request request, ActionListener<Response> li
     protected abstract Response newResponseInstance();
 
     /**
-     * Resolves the target shard id of the incoming request.
+     * Resolves derivative values in the request. For example, the target shard id of the incoming request,
+     * if not set at request construction
      * Additional processing or validation of the request should be done here.
+     *
+     * @param metaData cluster state metadata
+     * @param indexMetaData index metadata of the concrete index this request is going to operate on
+     * @param request the request to resolve
      */
-    protected void resolveRequest(MetaData metaData, String concreteIndex, Request request) {
-        // implementation should be provided if request shardID is not already resolved at request construction
+    protected void resolveRequest(MetaData metaData, IndexMetaData indexMetaData, Request request) {
```

Comment:
I've never been a fan of implementations where every derived class *must* call the super method is the case here. With this change, I would much prefer that `resolveRequest` be made final which delegates to a protected `doResolveRequest` that the derived classes override.